kernel/kernel-api/src/main/java/io/delta/kernel/internal/data/PaginationFilteredColumnarBatch.java

```diff
@@ -0,0 +1,53 @@
+package io.delta.kernel.internal.data;
+
+import static io.delta.kernel.internal.util.Preconditions.checkArgument;
+
+import io.delta.kernel.data.ColumnVector;
+import io.delta.kernel.data.ColumnarBatch;
+import io.delta.kernel.data.FilteredColumnarBatch;
+import java.util.Optional;
+
+/**
+ * An extension of {@link FilteredColumnarBatch} that adds metadata for pagination use, such as the
+ * originating file path and a pre-computed count of selected rows.
+ */
+public class PaginationFilteredColumnarBatch extends FilteredColumnarBatch {
+  private final String filePath;
+  private final int preComputedNumSelectedRows;
+
+  public PaginationFilteredColumnarBatch(
+      ColumnarBatch data,
+      Optional<ColumnVector> selectionVector,
+      String filePath,
+      int numSelectedRows) {
+    super(data, selectionVector);
+    this.filePath = filePath;
+    checkArgument(
```

Comment:
nit: I think the convention here would be to do the `checkArgument`s first and then do the `this.filePath = filepath` assignment

btw, great assertions!